File: AdisyonUygulamaJetpack/app/src/main/java/com/example/adisyonuygulamajetpack/viewmodel/UrunViewModel.kt
package com.example.adisyonuygulamajetpack.viewmodel

import android.util.Log
import androidx.lifecycle.*
import com.example.adisyonuygulamakotlin.model.Urun
import com.example.adisyonuygulamakotlin.model.Kategori
import com.example.qrkodolusturucu.services.ServicesImpl

import kotlinx.coroutines.launch

class UrunViewModel : ViewModel() {

    private val _urunler = MutableLiveData<List<Urun>>(emptyList())
    val urunler: LiveData<List<Urun>> = _urunler

    private val _kategoriler = MutableLiveData<List<Kategori>>(emptyList())
    val kategoriler: LiveData<List<Kategori>> = _kategoriler

    // Retrofit DAO
    private val dao = ServicesImpl.getInstance()

    fun urunleriYukle() {
        viewModelScope.launch {
            try {
                _urunler.value = dao.urunleriGetir()
            } catch (e: Exception) {
                Log.e("UrunVM", "Ürün yüklenemedi: ${e.localizedMessage}")
            }
        }
    }

    fun kategorileriYukle() {
        viewModelScope.launch {
            try {
                _kategoriler.value = dao.kategorileriGetir()
            } catch (e: Exception) {
                Log.e("UrunVM", "Kategori yüklenemedi: ${e.localizedMessage}")
            }
        }
    }

    fun urunEkle(ad: String, fiyat: Float, kategoriId: Int, base64: String,) {
        viewModelScope.launch {
            try {
                dao.urunEkle(ad, fiyat,kategoriId,0, base64)
            } catch (e: Exception) {
                Log.e("UrunVM", "Ürün ekleme hatası: ${e.localizedMessage}")
            }
        }
    }

    fun urunSil(urunAd: String) {
        viewModelScope.launch {
            try {
                dao.urunSil(urunAd)
            } catch (e: Exception) {
                Log.e("UrunVM", "Ürün silme hatası: ${e.localizedMessage}")
            }
        }
    }

    fun kategoriEkle(ad: String) {
        viewModelScope.launch {
            try {
                dao.kategoriEkle(ad)
            } catch (e: Exception) {
                Log.e("UrunVM", "Kategori ekleme hatası: ${e.localizedMessage}")
            }
        }
    }

    fun kategoriSil(kategoriId: Int) {
        viewModelScope.launch {
            try {
                dao.kategoriSil(kategoriId)
            } catch (e: Exception) {
                Log.e("UrunVM", "Kategori silme hatası: ${e.localizedMessage}")
            }
        }
    }

    companion object {
        val Factory = object : ViewModelProvider.Factory {
            override fun <T : ViewModel> create(modelClass: Class<T>): T {
                return UrunViewModel() as T
            }
        }
    }
}

File: flutter_adisyon_uygulama/lib/viewmodel/urun_viewmodel.dart
import 'package:flutter/foundation.dart';

import '../model/kategori.dart';
import '../model/urun.dart';
import '../service/services.dart';
import '../service/services_impl.dart';
import '../utils/extension.dart';

class UrunViewModel extends ChangeNotifier {
  final Services _dao;

  UrunViewModel({Services? dao}) : _dao = dao ?? ServicesImpl.getInstance();

  List<Urun> urunler = const [];
  List<Kategori> kategoriler = const [];

  bool loadingUrun = false;
  bool loadingKategori = false;

  Future<void> urunleriYukle() async {
    loadingUrun = true;
    notifyListeners();
    try {
      urunler = await _dao.urunleriGetir();
    } catch (e, st) {
      logE('UrunVM', e, st);
    } finally {
      loadingUrun = false;
      notifyListeners();
    }
  }

  Future<void> kategorileriYukle() async {
    loadingKategori = true;
    notifyListeners();
    try {
      kategoriler = await _dao.kategorileriGetir();
    } catch (e, st) {
      logE('UrunVM', e, st);
    } finally {
      loadingKategori = false;
      notifyListeners();
    }
  }

  Future<void> urunEkle({
    required String ad,
    required double fiyat,
    required int kategoriId,
    required String base64,
  }) async {
    try {
      await _dao.urunEkleAdmin(ad, fiyat, kategoriId, 0, base64);
      await urunleriYukle();
    } catch (e, st) {
      logE('UrunVM', e, st);
    }
  }

  Future<void> urunSil(String urunAd) async {
    try {
      await _dao.urunSil(urunAd);
      await urunleriYukle();
    } catch (e, st) {
      logE('UrunVM', e, st);
    }
  }

  Future<void> kategoriEkle(String ad) async {
    try {
      await _dao.kategoriEkle(ad);
      await kategorileriYukle();
    } catch (e, st) {
      logE('UrunVM', e, st);
    }
  }

  Future<void> kategoriSil(int kategoriId) async {
    try {
      await _dao.kategoriSil(kategoriId);
      await kategorileriYukle();
    } catch (e, st) {
      logE('UrunVM', e, st);
    }
  }
}
